### PR TITLE
0xMacro and ChainSecurity Audits Handling

### PR DESCRIPTION
**Pull request by rockyfour**

### CODE REVIEW

**Review by The-Arbiter — APPROVED** [review]
Changes look good. Removing 0 scope is a good move. My only concern was using uint32() as type for the file param, but this seems to be OK. The majority of other changes are formatting - removing the require around the approval in _swap should be OK since a failed approval will revert elsewhere.

**Comment by The-Arbiter** [comment]
I have gone through the responses of each auditor and evaluated each issue. I would appreciate further info on 6.4 if there is concern regarding that item.

My own feedback:
- I would edit the docs which describe `yen` as "The amount of acceptable slippage per lot" since a 100 * WAD `yen` value is not the slippage here. Editing the docs here would be useful since users may think that they want 0 slippage and set `yen` to 0 (which creates a host of other potential issues).
- We should elaborate on what `scope` refers to exactly. We say "The number of seconds to quote average price" which doesn't refer provide context for "the number of seconds" - "before now"? "before the swap"? A one liner here would be useful in my view; something like "A `scope` second window before before the quote is calculated is used to generate the 'average' value" (this may be wrong, just something to that effect). 

## ChainSecurity

6.1 - *Non-issue*
I understand their reasoning for raising this. Overall this is minor and I think sanitising the inputs in the file function is fine given that we are building this as more of a primitive that can work on its own rather than a dss component (which they compare it to).

6.2 - *Non-issue*
This is more of a user-error thing which is situation dependent. Again, if we are building it as a primitive then this feels more like something to go in the docs than to be enforced by code.

6.3 - *Non-functional issue*
Removing the `require` shouldn't cause substantial issues; the following call to
`SwapRouterLike(uniV3Router).exactInput(params)` should fail anyway if there is insufficient approved balance.

6.4 - *Functional issue (Unresolved)*
I don't fully get what was raised here but it feels like the min price value is recommended to be updated over time as the price increases. Assessment per the sheet is that updating this would complicate the code further, which we want to avoid. Someone else can chime in if this is a major issue but it doesn't feel serious.

## 0xMacro

M.1 - *Non-functional issue*
Sandwich attacks (manual, human performed) are possible because of the predetermined firing time of trades (calls to `fire` are permissionless). This has been discussed before and is very hard to avoid for large size TWAP orders regardless of the setup. Considering the optimal the size of each order to balance number of orders necessary with the size of each order would be best to mitigate this, but it cannot be fully prevented.

L.1 - *Functional issue (Unresolved)*
The issue of `buy` and `path` being decoupled causes as possible issue when users misconfigure a kiln instance. Code review should catch this but in any case, a warning in the docs would cover this at worst. Token being sold can only be sold if it matches with a token in path.

L.2 - *Functional issue (Unresolved)*
Having an event be emitted (especially `rug`, which would be uncommon to have to call if the kiln instance is working as intended) isn't a major issue. While it's possible that event watchers may react to this, at most it'd be an alert rather than a consequential action. I'm on the fence here about whether we want to add a require to support non-standard tokens (which as was raised, isn't something Maker would likely use) or just leave it as is. From a 'building a primitive' perspecitve I guess it would need fixing. Either way it's not an issue that can cause direct material losses in my opinion.

Q.1 - *Non-functional issue*
In the comments, 'lowering' and 'raising' were reversed. This has been rectified.

Q.2 - *Functional issue (Resolved)*
Scope could overflow and break `_swap`. This has been rectified. Funnily enough, ChainSecurity's issue of sanitising inputs in `file` could be considered negated due to the need to sanitise inputs for this. Having the `require` in `_swap` would be silly for gas reasons and I prefer the 'building a primitive' perspective on this, so sanitising in file (as has been done) is the best solution. 

Q.3 - *Functional issue (Unresolved)*
If `_min(GemLike(sell).balanceOf(address(this)), lot)` (the amount to sell in this firing) exceeds the `uint128` max value, it will break `fire`. This issue has not been rectified. Since we have `rug` built in, in my view this won't cause material losses due to loss of funds.

Q.4 - *Non-issue*
Illiquid UniV3 pools would probably not be used for kiln anyway and this is true for all TWAP trade mechanisms I think. As mentioned in our feedback, we can mitigate this by using kiln for pairs with good liquidity however we should mention this in the docs for if others want to set up kiln instances.

Q.5 - *Functional issue (Unresolved)*
Fire will revert if the sell token is not the first token in the path value - this is the same issue as L1 with a different effect (sell token not buy token) and therefore gets the same response of "code review should catch this". We also have `rug` so realistically there would be no material loss.

Q.6 - *Non-issue*
While a misconfigured kiln causing unfavourable swaps is possible, the `file` based configuration is standard across many `dss` contracts where the aforementioned 'unfavourable' outcomes can also happen. I would personally deem this a non-issue.

Q.7 - *Non-functional issue (Resolved)*
Documentation change, presumably 'greater than' changing to 'greater than or equal to'. This has been rectified.

I.1 - *Non-functional issue (Resolved)*
Low or zero `yen` (minimum proportion of quote price received per swap) could cause an unfavourable trade. I would consider the issue here to be the docs more than the code - this is more of a user error. This has been rectified.

I.2 - *Non-functional issue (Resolved)*
Zero (not low, must be zero) `scope` (time between trades) values could cause an unfavourable trade. Issue here pertains to flash-loan based attacks and therefore non-zero `scope` should be immune to this. Issue is rectified, however I think we may have gone too far when it comes to warning users not to use too low a `scope` value, unless there is another potential attack type which manifests for low (non-zero) `scope` values.

I.3 - *Functional issue (Unresolved)*
Rapid price crashes either require a low `yen` value (potentially dangerous) or could delay the `fire` execution. We have addressed this by noting that `yen` may need to change over time, however my approach here would be to set a suitable initial `yen` which doesn't need any maintenance. This issue is hard to fix given the volatility of certain tokens.

I.4 - *Functional issue (Unresolved)*
Rapid price increases with a low `yen` value could incentivise a sandwich attack. A static yen appears to cause some issues as raised in this issue and I3. A potential future solution could be a floating `yen` value with a floor value, however for now this is a hard problem to solve without resorting to "we will have to monitor and maintain it". Again, a suitable initial `yen` value would mitigate most material damage which could be incurred by both I3 and I4.

I.5 - *Functional issue (Unresolved)*
Multi-block oracle manipulation could cause low `scope` kiln instances to be manipulated. This would require substantial payoff which I feel would invoke an M1 style sandwich attack over multi-block manipulation, so I deem this to be a non-issue given low-swap-value usecases. I would deem this a low severity issue at most, echoing our feedback that we will monitor the state of multi-block oracle manipulation attacks and act accordingly.

File: src/KilnUniV3.sol [hunk 1]
```diff
@@ -101,7 +103,7 @@ contract KilnUniV3 is KilnBase, TwapProduct {
     }
 
     function _swap(uint256 amount) internal override returns (uint256 swapped) {
-        require(GemLike(sell).approve(uniV3Router, amount));
+        GemLike(sell).approve(uniV3Router, amount);
```
Comment (kmbarry1) [inline src/KilnUniV3.sol:106]: Although the `require` has been removed, there is still a subtle effect of the ERC20 interface in KilnBase.sol specifying a boolean return value for `approve`--Solidity will add a check that a value _is returned_, so tokens that don't return `true` on a successful approval will cause a revert here. I don't know offhand if any common tokens violate the standard in this way, but given that we support a non-standard `transfer`, we might as well support a non-standard `approve` (there's no security risk AFAICT). As a bonus, it should save a bit of gas.

Comment (rockyfour) [reply src/KilnUniV3.sol:106]: Nice catch. Handled in [7833018](https://github.com/makerdao/dss-kiln/pull/12/commits/7833018857da8ad4bbf9816188f58916497f0aea)

**Review by kmbarry1 — APPROVED** [review]
LGTM

The changes address the stated issues and do not seem to introduce any new problems. Caveat: I have not reviewed every aspect of this code but I don't think that's needed given the simple and isolated nature of changes.

**Pull request merged**